website/guide/programming.html

```diff
@@ -493,6 +493,32 @@ <h2 id="duktape-api">Duktape API</h2>
 duk_get_prop_string(ctx, -1, "my_state_variable");
 </pre>
 
+
+<h3>Hidden Symbol properties</h3>
+
+<p>If data needs to be associated with an object, but hidden from Ecmascript
+code, hidden <a href="#symbols">Symbols</a> can be used as a property key. The
+key is only accessible via the C API, unless passed to Ecmascript code from the
+C API. A common use case is to associated backing pointers/data to C memory.
+Symbols are created as a string, but with a specific prefix.</p>
```

Comment:
There's now a `DUK_HIDDEN_SYMBOL("my_context_data")` macro in the API which is preferred, so the "specific prefix" note is maybe unnecessary detail. Also the example should be updated to use the macro. The Symbols section (in master) discusses the macros.

Reply:
Yep, just saw the symbol macro changes after I submitted this PR. Will make the changes.

Will there be a `DUK_USER_HIDDEN_SYMBOL`, or maybe changing `DUK_HIDDEN_SYMBOL` to `DUK_INTERNAL_SYMBOL` to disambiguate a `"\xff" "\xff"` application style symbol from the duktape internal symbols? I would think it would be simpler to not have to explain the naming conventions in documentation (like uppercase vs. lowercase).

Reply:
Personally I'd prefer using "\xFFvalue" over "\xFF\xFFvalue" in calling code simply because it's shorter. It's difficult to enforce the case convention with a macro, but I don't think it has caused much confusion in practice.

No plan for a separate user hidden Symbol macro at present, but for it to work reliably as a macro it would either need to use the double \xFF prefix, or maybe a separate prefix byte could be used (e.g. 0x82 to 0xbf is still free) for user hidden Symbols.

In any case I think the Symbols section needs a short discussion of this; the details were moved to symbols.rst but maybe the very basics need to be in the Guide section too.

Reply:
Hmm, actually it would be easiest from a backwards compatibility perspective to:
- Reserve the 0xFF prefix for user hidden symbols entirely.
- Use 0x82 prefix for Duktape internal hidden symbols because no user code should be relying on the internal hidden symbols (they're not under versioning guarantees).

Reply:
I added the prefix byte change as a work item for #1120, I don't think I'll have time to make the change in 2.2.0.

Reply:
I'll open a pull for it - it's not a difficult change but I'm just tidying up odds and ends for 2.2.0. If it seems risk-free I can see if it can still get merged or not. But regardless of that, it would be a backwards compatible change, keeping DUK_HIDDEN_SYMBOL() but simply removing the constraints for user code.

Reply:
I was thinking just have the `DUK_USER_HIDDEN_SYMBOL` style macro just prepend the `"\xFF" "\xFF"` across the board, without worrying about the convention.

Would be simple enough maybe to make it into the 2.2.0?

Reply:
Yes, but I would not want to use such a macro myself :-)

Reply:
Ok, #1721 would make the 0x82 byte prefix change. It seems like a pretty small change so maybe I'll shoehorn it into 2.2.